server/src/main/java/org/elasticsearch/index/IndexMode.java

```diff
@@ -40,6 +51,45 @@ void validateWithOtherSettings(Map<Setting<?>, Object> settings) {
             }
         }
 
+        @Override
+        public void validateWithSource(DocumentParserContext context) {
+            validateTimestamp(context);
+        }
+
+        private void validateTimestamp(DocumentParserContext context) {
+            IndexableField[] fields = context.rootDoc().getFields(TIMESTAMP_FIELD);
+            if (fields.length == 0) {
+                throw new IllegalArgumentException("time series index @timestamp field is missing");
+            }
+
+            long numberOfValues = Arrays.stream(fields)
+                .filter(indexableField -> indexableField.fieldType().docValuesType() == DocValuesType.SORTED_NUMERIC)
+                .count();
+            if (numberOfValues > 1) {
+                throw new IllegalArgumentException("time series index @timestamp field encountered multiple values");
+            }
+
+            long timestamp = fields[0].numericValue().longValue();
+            if (context.mappingLookup().getMapper(TIMESTAMP_FIELD).typeName().equals(DateFieldMapper.DATE_NANOS_CONTENT_TYPE)) {
+                timestamp /= NSEC_PER_MSEC;
+            }
+
+            long startTime = context.indexSettings().getTimeSeriesStartTime();
+            long endTime = context.indexSettings().getTimeSeriesEndTime();
+
+            if (timestamp < startTime) {
+                throw new IllegalArgumentException(
+                    "time series index @timestamp value [" + timestamp + "] must be larger than " + startTime
+                );
+            }
+
+            if (timestamp > endTime) {
```

Comment:
I *think* we want end time *exclusive* here - so `>=` I think. Right @imotov ?

Reply:
@nik9000 that's a good question. We probably want one of them inclusive and another exclusive. If we want this to be consistent with histograms, then start time should be inclusive and end time exclusive - what do you think?